fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RangeValue.java

```diff
@@ -0,0 +1,442 @@
+import com.apple.foundationdb.record.query.plan.explain.ExplainTokensWithPrecedence;
+import com.apple.foundationdb.record.query.plan.plans.QueryResult;
+import com.apple.foundationdb.tuple.ByteArrayUtil2;
+import com.google.auto.service.AutoService;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A {@link StreamingValue} that is able to return a range that is defined as the following:
+ * <ul>
+ *     <li>an optional inclusive start (0L by default).</li>
+ *     <li>an exclusive end.</li>
+ *     <li>an optional step (1L by default).</li>
+ * </ul>
+ * For more information, see relational SQL {@code range} table-valued function.
+ */
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+public class RangeValue extends AbstractValue implements StreamingValue, CreatesDynamicTypesValue {
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("Range-Value");
+
+    @Nonnull
+    private final Value endExclusive;
+
+    @Nonnull
+    private final Optional<Value> beginInclusive;
+
+    @Nonnull
+    private final Optional<Value> step;
+
+    @Nonnull
+    private final Value currentRangeValue;
+
+    public RangeValue(@Nonnull final Value endExclusive, @Nonnull final Optional<Value> beginInclusive,
+                      @Nonnull final Optional<Value> step) {
+        this.endExclusive = endExclusive;
+        this.beginInclusive = beginInclusive;
+        this.step = step;
+        currentRangeValue = RecordConstructorValue.ofColumns(ImmutableList.of(Column.of(Optional.of("ID"), LiteralValue.ofScalar(-1L))));
+    }
+
+    @Nonnull
+    @Override
+    public Type.Record getResultType() {
+        return (Type.Record)currentRangeValue.getResultType();
+    }
+
+    @Nonnull
+    @Override
+    public <M extends Message> RecordCursor<QueryResult> evalAsStream(@Nonnull final FDBRecordStoreBase<M> store,
+                                                                      @Nonnull final EvaluationContext context,
+                                                                      @Nullable final byte[] continuation,
+                                                                      @Nonnull final ExecuteProperties executeProperties) {
+        final long endExclusiveValue = (Long)Verify.verifyNotNull(endExclusive.eval(store, context));
+        final var beginInclusiveMaybe = beginInclusive.map(b -> (Long)Verify.verifyNotNull(b.eval(store, context)));
```

Comment:
In the planner I adopted for methods that return `Optional<>` to be named `...Maybe`, variables that are `Optional` are called `...Optional`. I think I picked that up from some framework in Scala. It's debatable if we want that, but we should not deviate from the rule unless we change it everywhere. We the other suggested change this should go away in any case (I hope)...

Reply:
No optionals anymore.